File: leetcode/15/main.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  const sorted_nums = [...nums].sort((a, b) => a - b);

  const ret = [];

  const N = sorted_nums.length;
  for (let i = 0; i + 2 < N; ++i) {
    if (i > 0 && sorted_nums[i] === sorted_nums[i - 1]) continue;
    if (sorted_nums[i] > 0) break;

    let l = i + 1,
      r = N - 1;

    while (l < r) {
      const sum = sorted_nums[i] + sorted_nums[l] + sorted_nums[r];
      if (sum < 0) l++;
      else if (sum > 0) r--;
      else {
        ret.push([sorted_nums[i], sorted_nums[l], sorted_nums[r]]);
        (l++, r--);
        while (l < r && sorted_nums[l] == sorted_nums[l - 1]) l++;
        while (l < r && sorted_nums[r] == sorted_nums[r + 1]) r--;
      }
    }
  }

  return ret;
};

File: leetcode/15/main.cpp
#include <vector>
using namespace std;

class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    sort(nums.begin(), nums.end());

    vector<vector<int>> ret;
    for (int i = 0; i + 2 < nums.size(); ++i) {
      if (i > 0 && nums[i] == nums[i - 1]) continue;
      if (nums[i] > 0) break;

      int l = i + 1, r = nums.size() - 1;
      while (l < r) {
        int sum = nums[i] + nums[l] + nums[r];
        if (sum == 0) {
          ret.push_back(vector<int>({ nums[i], nums[l++], nums[r--] }));
          while (l < r && nums[l] == nums[l - 1]) { l++; }
          while (l < r && nums[r] == nums[r + 1]) { r--; }
        } else if (sum < 0) {
          l++;
        } else {
          r--;
        }
      }
    }

    return ret;
  }
};
